### ISSUE TITLE
use_self triggers on code inside macros

### ISSUE THREAD

**Issue by jacobhenn** [rust-lang/rust-clippy#9786]
### Summary

the lint `clippy::use_self` will emit a false positive on code generated by a macro which triggers it.

### Lint Name

use_self

### Reproducer

this minimal example:

```rust
use derive_more::Display;

#[derive(Display)]
struct Foo;
```

gives this output:

```
warning: unnecessary structure name repetition
  --> src/main.rs:8:8
   |
 8 | struct Foo;
   |        ^^^ help: use the applicable keyword: `Self`
   |
   = note: `-W clippy::use-self` implied by `-W clippy::nursery`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
```

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by kraktus** [comment]
should be fixed by https://github.com/rust-lang/rust-clippy/pull/9454, can you reproduce on nightly?

**Comment by jacobhenn** [comment]
yeah, that fixes it. sorry, i tried to search for a pre-existing issue but those didn't come up.

**Comment by kraktus** [comment]
No worries thanks for reporting!
